### ISSUE TITLE
document.execCommand("copy") does not work in cypress

### ISSUE THREAD

**Issue by tnrich** [cypress-io/cypress#2851]
Hi there,

I'm curious as to why a `document.execCommand("copy")` fails when being run during a cypress test (as a side effect to a cy.get().click() ). Curiously it works fine once I'm manually clicking the same element after the test has run. 

The failing test in question can be reproduced by running the following code: 
```
describe("menuBar", function() {
  beforeEach(() => {
    cy.visit("http://teselagen.github.io/openVectorEditor/#/Editor");
  });
  
  it("select range, copy, cut works", function() {
    cy.clock()
    cy.get('.tg-menu-bar').contains("Edit").click()
    cy.get('.tg-menu-bar-popover').contains("Select").click()
    cy.get(`[label="From:"]`).clear().type("10")
    cy.get(`[label="To:"]`).clear().type("20")
    cy.get(`.dialog-buttons`).contains("OK").click()
    cy.get(".veStatusBar").contains(`10 to 20`)

    cy.get(".veStatusBar").contains(`5299`)
    cy.get('.tg-menu-bar').contains("Edit").click().tick(200)
    cy.get('.tg-menu-bar-popover').contains("Copy").click()
    cy.contains("Selection Copied")
    cy.get('.tg-menu-bar').contains("Edit").click()
    cy.get('.tg-menu-bar-popover').contains("Cut").click()
    cy.contains("Selection Cut")
    cy.get(".veStatusBar").contains(`5288`)
  });
});
```

If you run the above example, the test will fail to find "Selection Copied". This is because 
```
const worked = document.execCommand(type);
```
evaluates to false so the "Selection Copied" popover is not shown. 

You can add a breakpoint manually to the `commands/index.js` file and see it not working. 
![image](https://user-images.githubusercontent.com/2730609/49178000-7baa2600-f303-11e8-8e48-df7449b81094.png)

Strangely, after the test has run if you manually click "File > Edit > Copy" then the document.execCommand("copy") works just fine. Not sure what is going on here to stop it from working when cypress is running its tests. 

![image](https://user-images.githubusercontent.com/2730609/49178108-c1ff8500-f303-11e8-9c9f-c8325a974cc9.png)

![image](https://user-images.githubusercontent.com/2730609/49178143-d80d4580-f303-11e8-97b0-8d30dd889726.png)

Thanks!

**Comment by jennifer-shehane** [comment]
I thought this piece of info from MDN docs looks interesting concerning the return value of `document.execCommand()`

>Note: Only returns `true` if part of a user interaction. Don't try using the return value to verify browser support before calling a command. 

Since Cypress programmatically triggers the 'click' event, the `document.execCommand()` will always return true in this case. This looks like another reason we are implementing native events. https://github.com/cypress-io/cypress/issues/311

**Comment by tnrich** [comment]
@jennifer-shehane so in the future I should be able to do cy.get().click({native:true}) to have the test pass as expected? Is there an expected date the `native` events feature will land? 

Thanks!

**Comment by tnrich** [comment]
@Bkucera passing true/false as the second arg doesn't seem to have any effect on the copy command getting triggered from the cypress click.

**Comment by Bkucera** [comment]
@jennifer-shehane is correct, the copy command can only happen in response to a _trusted_ user event, which can only be done with native events.
I'll add this to the native events issue

**Comment by jennifer-shehane** [comment]
Related to 
- Using modifier keys for copy + paste: https://github.com/cypress-io/cypress/issues/2386
- Enable access to clipboard contents: https://github.com/cypress-io/cypress/issues/2752

**Comment by asherccohen** [comment]
@jennifer-shehane  Any news on copy/pasting feature in Cypress?

**Comment by bhellema** [comment]
@jennifer-shehane we're also looking for an update on this.  Thanks

**Comment by xcao-riffyn** [comment]
agreed with @bhellema really hoping this eventually will work.  I might be able to offer some help.  The biggest barrier that I can see is how cypress sends key combinations.  I know an npm module that could copy paste the content from current clipboard: https://www.npmjs.com/package/copy-paste maybe we can incorporate that into cypress

**Comment by dwelle** [comment]
@xcao-riffyn copying to/from clipboard in cypress [isn't a problem](https://github.com/cypress-io/cypress/issues/2752#issuecomment-476279234). This issue is about `document.execCommand("copy")` not working and using a lib to manipulate OS clipboard isn't gonna solve it because we need to test AUT behavior --- our app itself uses `document.execCommand("copy")` so that's why we need to make that DOM command work properly.

**Comment by Bkucera** [comment]
There is a way we can make this testable. We can overwrite `document.execCommand` and store the copied text in a "fake" clipboard object that exists during the lifetime of the test. That way you could do something like:
```js
cy.get('button').click().then(()=>{
  cy.clipboard().should('contain', 'sometext')
})
```

This is just one possible version of the API. I would appreciate feedback

**Comment by dwelle** [comment]
@Bkucera problem with mocking is you won't know if `document.execCommand` will succeed in real conditions. As an interim solution, sounds good.

**Comment by Bkucera** [comment]
@dwelle I believe we can check the stack for a cypress click command, using rules that mimic the rules that the browser follows for user interaction.

**Comment by asherccohen** [comment]
Is there any update on this?

**Comment by tnrich** [comment]
@Bkucera ship it! I'd love to use your suggested 
```
cy.get('button').click().then(()=>{
  expect(cy.clipboard()).to.contain('sometext')
})
```

**Comment by rahul-sharma-uipath** [comment]
any update on this?

**Comment by jennifer-shehane** [comment]
This issue is still in the 'ready for work' stage, which means no work has been done on this issue as of today, so we do not have an estimate on when this will be delivered.

**Comment by agugut** [comment]
Is there any update on this?

**Comment by bahmutov** [comment]
Initial work to show how to access the clipboard from the test is in this recipe https://github.com/cypress-io/cypress-example-recipes/tree/master/examples/testing-dom__clipboard

In general, applications using `navigator.clipboard` API seem to work from Cypress in Chrome and Electron browsers. The applications using `document.execCommand` seem to work in Electron browser only.

In your particular case I ran the test as shown below in Electron using Cypress v7.4.0 and it passed

```js
/// <reference types="cypress" />
describe("menuBar", function() {
  beforeEach(() => {
    cy.visit("/openVectorEditor/#/Editor");
  });

  it("select range, copy, cut works", function() {
    cy.clock()
    cy.get('.tg-menu-bar').contains("Edit").click()
    cy.get('.tg-menu-bar-popover').contains("Select").click()
    cy.get(`[label="From:"]`).clear().type("10")
    cy.get(`[label="To:"]`).clear().type("20")
    cy.get(`.dialog-buttons`).find('[type=submit]').click()
    cy.get(".veStatusBar").contains(`10 to 20`)

    cy.get(".veStatusBar").contains(`5299`)
    cy.get('.tg-menu-bar').contains("Edit").click().tick(200)
    cy.get('.tg-menu-bar-popover').contains("Copy").click()
    cy.contains("Selection Copied")
    cy.get('.tg-menu-bar').contains("Edit").click()
    cy.get('.tg-menu-bar-popover').contains("Cut").click()
    cy.contains("Selection Cut")
    cy.get(".veStatusBar").contains(`5288`)
  });
});
```

<img width="1276" alt="Screen Shot 2021-05-26 at 3 43 39 PM" src="https://user-images.githubusercontent.com/2212006/119721540-3f56d280-be39-11eb-93cb-f3da520eefac.png">

**Comment by cx-ruan-carvalho** [comment]
Any news on it? There is some bypass to this?

**Comment by Bjyotika** [comment]
Any update on this as due to this issue unable to test  in cypress which require clipboard copy approach

**Comment by MattSzCosta** [comment]
Hey guys, I had the same issue while testing copy/paste behavior on the rich text editor. I managed to work around the problem. It's a bit tricky, but it worked for me.

I add an eventListnner to the copy event and send it to a callback function to manipulate the data.
```js
Cypress.Commands.add('copy', { prevSubject: true }, (subject, onCopy) => {
  cy.window().then((win) => {
    cy.stub(win.document, 'execCommand').callsFake(() => true)
    win.document.addEventListener('copy', (event) => {
      onCopy?.(event)
    })
  })
  const clipboardData = new DataTransfer()
  const copyEvent = new ClipboardEvent('copy', {
    bubbles: true,
    cancelable: true,
    clipboardData,
  })
  subject[0].dispatchEvent(copyEvent)
})
```

Usage:
```js
const onCopy = (event: ClipboardEvent) => {
  const copiedHtmlText = event.clipboardData?.getData('text/html') || ''
  expect(copiedHtmlText).contains(`input text`)
}

cy.findByLabelText('.input').then(($editable) => {
  const editable = $editable[0]
  const range = document.createRange()
  range.selectNodeContents(editable)
  const selection = window.getSelection()
  selection?.removeAllRanges()
  selection?.addRange(range)

  cy.wrap(editable).copy(onCopy)
})
```

```js
Cypress.Commands.add('paste', { prevSubject: true }, (subject, payload) => {
  const clipboardData = new DataTransfer()
  Object.keys(payload).forEach((dataType) => {
    clipboardData.setData(dataType, payload[dataType])
  })
  const pasteEvent = new ClipboardEvent('paste', {
    bubbles: true,
    cancelable: true,
    clipboardData,
  })
  subject[0].dispatchEvent(pasteEvent)
  return subject
})

// Usage
cy.findAllByTestId('.input').paste({ 'text/html': "<p>test</p>", 'text/plain': "test" })
```
